go/teams/transactions.go

```diff
@@ -0,0 +1,495 @@
+	return nil
+}
+
+func (tx *AddMemberTx) Post(ctx context.Context) (err error) {
+	if len(tx.payloads) == 0 {
+		return errors.New("there are no signatures to post")
+	}
+
+	team := tx.team
+	g := team.G()
+
+	defer g.CTrace(ctx, "AddMemberTx.Post", func() error { return err })()
+	g.Log.CDebugf(ctx, "AddMemberTx: Attempting to post %d signatures", len(tx.payloads))
+
+	// Initialize key manager.
+	if _, err := team.SharedSecret(ctx); err != nil {
+		return err
+	}
+
+	// Make sure we know recent merkle root.
+	if err := team.ForceMerkleRootUpdate(ctx); err != nil {
+		return err
+	}
+
+	// Get admin permission, we will use the same one for all sigs.
+	admin, err := team.getAdminPermission(ctx, true)
+	if err != nil {
+		return err
+	}
+
+	var sections []SCTeamSection
+	memSet := newMemberSet()
+
+	// Transform payloads to SCTeamSections.
+	for _, p := range tx.payloads {
+		section := SCTeamSection{
+			ID:       SCTeamID(team.ID),
+			Admin:    admin,
+			Implicit: team.IsImplicit(),
+			Public:   team.IsPublic(),
+		}
+
+		switch payload := p.(type) {
+		case *keybase1.TeamChangeReq:
+			// We need memberSet for this particular payload, but also keep a
+			// memberSet for entire transaction to generate boxes afterwards.
+			payloadMemberSet, err := newMemberSetChange(ctx, g, *payload)
+			if err != nil {
+				return err
+			}
+
+			// TODO: Instead of loading members twice, have a "append"
+			// function in memberSet.
+			if err := memSet.loadMembers(ctx, g, *payload, true /* forcePoll */); err != nil {
+				return err
+			}
+
+			section.Members, err = payloadMemberSet.Section()
+			if err != nil {
+				return err
+			}
+
+			section.CompletedInvites = payload.CompletedInvites
+			sections = append(sections, section)
+		case *SCTeamInvites:
+			entropy, err := makeSCTeamEntropy()
+			if err != nil {
+				return err
+			}
+
+			section.Invites = payload
+			section.Entropy = entropy
+			sections = append(sections, section)
+		default:
+			return fmt.Errorf("Unhandled case in AddMemberTx.Post, unknown type: %T", p)
+		}
+	}
+
+	// If memSet has any downgrades, request downgrade lease.
+	var merkleRoot *libkb.MerkleRoot
+	var lease *libkb.Lease
+
+	downgrades, err := team.getDowngradedUsers(ctx, memSet)
+	if len(downgrades) != 0 {
+		lease, merkleRoot, err = libkb.RequestDowngradeLeaseByTeam(ctx, g, team.ID, downgrades)
+		if err != nil {
+			return err
+		}
+		// Always cancel lease so we don't leave any hanging.
+		defer func() {
+			err := libkb.CancelDowngradeLease(ctx, g, lease.LeaseID)
+			if err != nil {
+				g.Log.CWarningf(ctx, "Failed to cancel downgrade lease: %s", err.Error())
+			}
+		}()
+	}
+
+	secretBoxes, implicitAdminBoxes, perTeamKeySection, err := team.recipientBoxes(ctx, memSet)
+	if err != nil {
+		return err
+	}
+
+	if perTeamKeySection != nil {
+		// We have a new per team key, find first TeamChangeReq
+		// section created and add perTeamKeySection.
+		// TODO: Should it be the first ChangeMembership sig or the last?
```

Comment:
Either way does the set of people who have access to it reflect the _final_ set of members? Then I'd think it should be the last.

Reply:
I think this will not work either way if there is a transaction that has more than one ChangeMembership. If it's in the first, then server should reject the sig because of over-keying, if it's in the last, server could reject prior sigs because of under-keying.

Reply:
Does the server really check each sig or does it run a keying invariant check at the end of the transaction?